waggle-dance-core/src/test/java/com/hotels/bdp/waggledance/mapping/service/PanopticConcurrentOperationExecutorTest.java

```diff
@@ -0,0 +1,138 @@
+  }
+
+  @Test
+  public void executeRequestsInOrder() throws Exception {
+    PanopticConcurrentOperationExecutor executor = new PanopticConcurrentOperationExecutor();
+    List<DummyRequestCallable> allRequests = Lists
+        .newArrayList(new DummyRequestCallable("call1", mapping1), new DummyRequestCallable("call2", mapping2));
+    List<String> executeRequests = executor.executeRequests(allRequests, REQUEST_TIMEOUT, "error");
+    assertThat(executeRequests.size(), is(2));
+    assertThat(executeRequests.get(0), is("call1"));
+    assertThat(executeRequests.get(1), is("call2"));
+  }
+
+  @Test
+  public void executeRequestsSlowConnection() throws Exception {
+    PanopticConcurrentOperationExecutor executor = new PanopticConcurrentOperationExecutor();
+    DummyRequestCallable slowRequest = new DummyRequestCallable("call2", mapping2) {
+      @Override
+      public List<String> call() throws Exception {
+        // Too slow should never be executed
+        Thread.sleep(REQUEST_TIMEOUT * 10);
+        return super.call();
+      };
+    };
+    List<DummyRequestCallable> allRequests = Lists
+        .newArrayList(new DummyRequestCallable("call1", mapping1), slowRequest);
+    List<String> executeRequests = executor.executeRequests(allRequests, REQUEST_TIMEOUT, "error");
+    assertThat(executeRequests.size(), is(1));
+    assertThat(executeRequests.get(0), is("call1"));
+  }
+
+  @Test
+  public void executeRequestsWaitForMaxLatency() throws Exception {
+    Long mapping2Latency = TimeUnit.MILLISECONDS.toMillis(1000);
+    when(mapping2.getLatency()).thenReturn(mapping2Latency);
+    PanopticConcurrentOperationExecutor executor = new PanopticConcurrentOperationExecutor();
+    DummyRequestCallable slowRequest = new DummyRequestCallable("call2", mapping2) {
+      @Override
+      public List<String> call() throws Exception {
+        // Slow but within timeout+latency, should be called
+        Thread.sleep(REQUEST_TIMEOUT + 100);
+        return super.call();
+      };
+    };
+    List<DummyRequestCallable> allRequests = Lists
+        .newArrayList(new DummyRequestCallable("call1", mapping1), slowRequest);
+    List<String> executeRequests = executor.executeRequests(allRequests, REQUEST_TIMEOUT, "error");
+    assertThat(executeRequests.size(), is(2));
+    assertThat(executeRequests.get(0), is("call1"));
+    assertThat(executeRequests.get(1), is("call2"));
+  }
+
+  @Test
+  public void executeRequestsExceptionLoggedResultsReturned() throws Exception {
+    PanopticConcurrentOperationExecutor executor = new PanopticConcurrentOperationExecutor();
+    DummyRequestCallable errorRequest = new DummyRequestCallable("call2", mapping2) {
+      @Override
+      public List<String> call() throws Exception {
+        throw new RuntimeException("who you gonna call...");
+      };
+    };
+    List<DummyRequestCallable> allRequests = Lists
+        .newArrayList(new DummyRequestCallable("call1", mapping1), errorRequest);
```

Comment:
What if we make `call1` return an error instead of `call2`? This way we can check that 1 failed request won't affect all of them.

Reply:
calls are done concurrently right? how is that different? I'll add a call3 to cover all bases, makes it more clear

Reply:
Oh, actually you're right. They are all resolved at the same time with `invokeAll`. This should be fine then 😄